### ISSUE TITLE
Bug: [@typescript-eslint/no-base-to-string] False positive on * import containing toString function with overloads

### ISSUE THREAD

**Issue by bschlenk** [typescript-eslint/typescript-eslint#11945]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

If you `*` import everything from a module, and that module contains a `toString` method that has function overloads, then the `@typescript-eslint/no-base-to-string` rule incorrectly reports that the `toString` call is going to use the default implementation.

The function overloads appear to be an important part of the repro. Without them the rule doesn't report anything.

### Reproduction Repository Link

https://github.com/bschlenk/no-base-to-string-repro

### Repro Steps

1) clone the repo
2) `pnpm install`
3) `pnpm lint`

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `typescript-eslint`     | `8.52.0` |
| `TypeScript`                           | `5.9.3` |
| `ESLint`                               | `9.39.2` |
| `node`                                 | `22.17.1` |

**Comment by JoshuaKGoldberg** [comment]
😂 I have to know: how did you discover this? I have never seen a project rely on using a module like an object with a `toString()`.

**Comment by bschlenk** [comment]
I agree it's a bit odd! My company's code is in one giant mono repo and we sometimes try to enforce all packages across the mono repo to use the same version of dependencies. So when I bumped all the ts eslint dependences from 8.47.0 to 8.52.0, it just happened to start catching code like this.

As for _why_ the code is like this, we have a concept of a GUID that has `sessionId` and `localId` components. For perf reasons, we keep it as 2 numbers when we pass them between wasm and js. But often we want to see them as a nice readable string, so the `toString` function just joins those with a `:`. And often all the GUID related utils are imported as `import * as GUID from './guid'`, which pulls in the `toString` method. In this case we call `toString` with the actual guid as the argument. Not trying to get a string representation of the module itself.

**Comment by HolgerJeromin** [comment]
I found the same false positive.
We have a `ValueConverter` module which converts user API input to `Number`, `String`, `Boolean`. 
Part of it is the `toString` method with many overloads for nicer intellisense.
```ts
export function toString<T extends string = string>(value: T): T | null; // support ValueConverter.toString('foo');
export function toString<T extends string | null | undefined = string>(value: any | null | undefined): NonNullable<T> | null; // support ValueConverter.toString<'foo'|null>('foo');
export function toString(value: any): string | null; // support ValueConverter.toString({'foo':true});
// ...
```
used like this
```ts
import * as ValueConverter from '../../../API/ValueConverter.js';
// ...
let convertedValue = ValueConverter.toString(valueNew);
```

**Comment by kirkwaiblinger** [comment]
What an odd edge case 😅 

I suspect this is related to the hacky `declarations?.length` check in here: 

https://github.com/typescript-eslint/typescript-eslint/blob/309a38ed83994738323efd78fc31137136a7681a/packages/eslint-plugin/src/rules/no-base-to-string.ts#L349-L375

This was originally added in https://github.com/typescript-eslint/typescript-eslint/issues/8585 -> https://github.com/typescript-eslint/typescript-eslint/pull/10432 as a best effort heuristic for detecting user-defined `toString()` methods, but probably has some kind of mistake 🤷 

FWIW, this is all a little tricky because IIRC an `import * as Foo` creates a null-prototype object that wouldn't have a `.toString()` method at all normally. So the usual assumption that finding exactly one declaration for `.toString()` means it could be the built-in `.toString()` might not be appropriate here. Needs some poking around

**Comment by renkei** [comment]
This bug is not limited to multiple toString() overloads from a module imported via import * as xyz syntax. Regular classes and even interfaces defined for classes or objects are also affected if you provide multiple overloads.

For example, I implemented a single toString(prefix?: string) method in my class which worked fine. Later, I needed more flexibility for more control and therefore added a toString(options: StringOptions) overload. It includes the first overload, but for backward compatibility reasons, I left the first overload in my class. Worked fine until @typescript-eslint 8.51 and is false positive since 8.52 and newer.

Of cource, I can merge both signatures into a single one, like toString(options?: string | StringOptions), this would solve the issue and additionally passes the @typescript-eslint/unified-signatures rule. I just wanted to mention that multiple overloads are generally problematic for this rule in version ≥ 8.52.

**Comment by HolgerJeromin** [comment]
Just a note: 
```js
'@typescript-eslint/no-base-to-string': ['error', {  ignoredTypeNames: ['ValueConverter'], }]
``` 
does not fix the error in my case
```ts
import * as ValueConverter from '../../../API/ValueConverter.js';
```

**Comment by cakoose** [comment]
Here's another very simple repro ([playground link](https://typescript-eslint.io/play/#ts=5.9.3&fileType=.tsx&code=CYUwxgNghgTiAEkoGdnwEIpAHgCoD54BvAKHngBcB7AZQpgEsA7AcwAoBKALnmXuZYBuMvAD0o%2BAEkAZvACeVAK6IqAW1UgmFALRKKlABYIARiAhUA7vAjMQAGkMIQMGFRjwWVEGigWocgDoRajpGVjZpN1UoCh4%2BMJZuXn5WYQBfEhJQJDhEaFR4AGEDBghgeBAADwpNYDRMZBxFJgBrJksmQiJ4DJJpZrAKBiomeGk2MB5i0uAOYhEwEeQqCBAA83YwAJCU9g4OQTEJaSgIRvgABypkBiGANxAAQhIMoA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oCMBDZRLXxdk%2BKkwDm6KImjQO0SODABfECqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false)):

```ts
declare class Base<T> {
  toString(): string;
  // If you comment-out the below line, the error goes away.
  toString(format: string): string;
}

declare class Child extends Base<unknown> { }

function f(c: Child) {
  console.log(c.toString()); // false positive!
}
```

TSESLint 8.56.1.
- We weren't getting this error in 8.49.0.
